spec/services/slack_service/group_member_messaging_spec.rb

```diff
@@ -115,4 +115,41 @@
       expect(list_block[:text]).to include("20 minutes")
     end
   end
+
+  describe "_format_desc" do
+    let(:entry_1)   { create(:entry, description: "this is an entry", minutes: 120) }
+    let(:entry_2)   { create(:entry, description: "doing something #unbillable*", minutes: 90) }
+    let(:entry_3)   { create(:entry, description: "working on #pecas  #unbillable*  #calls", minutes: 20) }
+    let(:entry_4)   { create(:entry, description: "testing this entry with a dash #call-pecas  #unbillable*  #calls", minutes: 20) }
+
+    before do
+      allow(Slack::Web::Client).to receive(:new).and_return(client)
+      allow(SlackService).to receive(:find_usergroup_id).with("avengers", client).and_return([:ok, "GROUPID"])
+      allow(SlackService).to receive(:find_usergroup_user_ids).with("GROUPID", client).and_return([:ok, ["USERID1", "USERID2"]])
+      allow(SlackService).to receive(:find_slack_user).with("USERID1", client).and_return([:ok, user_1])
+      allow(SlackService).to receive(:find_slack_user).with("USERID2", client).and_return([:ok, user_2])
+      @service = SlackService::GroupMemberMessaging.new("avengers", six_oclock_hour, now)
+      @message = @service.send :time_entry_format_warning_message, user_1, [entry_1, entry_2, entry_3, entry_4]
+      @formatted_list = @message[:blocks][1][:text][:text].split("\n")
```

Comment:
I think the setup is too complex for what this is actually testing, I think this is a sign of something that can be refactored to make the test easier:

My suggestion would be:
- Extract the `format_desc` method as a class method, it doesn't really depend on the state of the `GroupMemberMessaging` object
- Extract the content of the entries loop as `format_entry` also making it a class method

that way, to test the formatting you don't need to creat all the extra stubbing for the SlackService and GroupMemberMessaging objects, you can just do `expect(SlackService.format_entry(entry1)).to eql(....)`, for each entry

By doing that you can also skip the `create` of the entries and just use `build`, so we don't have to write to the DB, making the tests faster, we don't really need to store these things since the method we are testing doesn't care about them being saved, it just cares about the desription and the length

Reply:
also, using `let` naming things like `entry_1/2/3/4` that are only used once is a sign that they don't have to be added as `let`s, you could do something like

```rb
    it "formats an entry with multiple labels" do
      entry = build(:entry, description: "working on #pecas #unbillable* #calls", minutes: 20)
      formatted = SlackService::GroupMemeberMessaging.format_entry(entry)
      expect(formatted).to eql("* working on `#pecas`   `#unbillable*`   `#calls`  (20 minutes)")
    end
```

That way, you have a more readable test, without having to look for what the tests involves in other places of the file:
- you have a setup
- you execute the method you are testing with that setup
- you make an assertion on the result

it's all there together